### DISCUSSION TITLE
Approaches to Sentry.io integration for capturing and surfacing errors

### DISCUSSION THREAD

**Discussion by horatiorosa** [NYCPlanning/ae-zoning-api#203] [Ideas]
#### Pull Request
[nestjs-sentry implementation](https://github.com/NYCPlanning/ae-zoning-api/pull/197)

####  Repo Stats:
- 365 stars
- 90 forks
- 39 open issues
- 7 open PRs

### Description:
@ntegral/nestjs-sentry implements a module, SentryModule, which when imported into your nestjs project provides a Sentry.io client to any class that injects it. This lets Sentry.io be worked into your dependency injection workflow without having to do any extra work outside of the initial setup.

### Setup:
- install npm package for nestjs-sentry
- Sentry init is in app.module.ts

####  Pros:
- Decent docs, docs come with sample implementation
- Set up and configeration is straightforward, only need to install one package
- Supports nestjs 10
- Can instruct Nest to use the SentryService as the default logger (lots of detail)
- Can instruct use the SentryService to log non-error messages as breadcrumbs or use the native Sentry.addBreadcrumb method
- Supports using the sentry interceptor globally
- Making a good effort to support current Nestjs versions
####  Cons:
- Seemingly infrequent updates
  - hardydex (top contrib): 2 yrs ago 
  - ntergral (code owner): 3 months ago, before that 2 years ago
- No releases / tags
 - Not as many contributors

#### Pull Request
[nest-racen implementation](https://github.com/NYCPlanning/ae-zoning-api/pull/201)
####  Repo Stats:
- 241 stars
- 25 forks
- 5 open PRs (all dependabots)
- 8 issues

####  Description:
This is a sentry module for Nest. 
This is not a solution for all cases and large applications, just a quick starter that covers the common rest/graphql capturing (the basics). You might want deeper integration with Sentry, which means you can still use this to cover the rest/graphql error capture, but you will have to use the rest of Sentry SDK to cover other cases.

For really large projects, you might have to take this library as an example of how to integrate sentry with NestJS, and write your custom integration instead.

####  Setup:
- Must install and set up Sentry as per Sentry docs, then install and set up nest-raven 
- Sentry init is in mail.ts (earliest entry point as recommended by Sentry)

####  Pros:
- Recent active contributor (9renpoto) somewhat active
- Has releases / versions
- Set up is fairly straightforward
####  Cons:
- Need to set up sentry SDK yourself via sentry's docs then set up nest-raven module
- Docs are a bit light
- May be a problem if the Zoning API gets bigger as sentry-raven isn't suitable for large applications
- Repo owner (mentos1386) last big commit was 2 years ago

###  For Further Discussion:

How  do we want to control the level of transaction sampling? 
- A general percentage (for example: tracesSampleRate: 0.2 // 20%) 
- Alternatively, we can capture varying percentages depending on the situation (like below):
```
tracesSampler: samplingContext => {
    // Examine provided context data (including parent decision, if any) along
    // with anything in the global namespace to compute the sample rate or
    // sampling decision for this transaction

    if ("...") {
      // These are important - take a big sample
      return 0.5;
    } else if ("...") {
      // These are less important or happen much more frequently - only take 1%
      return 0.01;
    } else if ("...") {
      // These aren't something worth tracking - drop all transactions like this
      return 0;
    } else {
      // Default sample rate
      return 0.1;
    }
  };
  ```

**Comment by TylerMatteo** [comment]
@horatiorosa Great job on this write up! Let's go ahead with `@ntegral/nestjs-sentry` for now. It seems to have better documentation and easier set up. Also, it seems to have a company (ntegral) using it internally so I feel pretty good about it being maintained, despite having relatively infrequent updates on paper.
